index.bs

```diff
@@ -1840,16 +1530,104 @@ partial interface MLGraphBuilder {
+  <div class=algorithm-steps>
+    <div class="note">
+        The permissions and context validity have been checked by [[#api-mlgraphbuilder-constructor]] steps.
+    </div>
+    1. Let |promise| be [=a new promise=].
+    1. Return |promise| and run the following steps [=in parallel=].
+    1. Return the result of invoking {{MLGraphBuilder/buildSync(outputs)}} given |outputs|.
+        1. If that [=exception/throws=], re-[=exception/throw=] the error.
+  </div>
+</details>
+
+#### {{MLGraphBuilder/buildSync(outputs)}} #### {#api-mlgraphbuilder-buildsync-outputs}
+
+<details open algorithm>
+  <summary>
+    The <dfn method for=MLGraphBuilder>buildSync(|outputs|)</dfn> method steps are:
+  </summary>
+  <div class=algorithm-steps>
+    <div class="note">
+        The permissions and context validity have been checked by [[#api-mlgraphbuilder-constructor]] steps.
+    </div>
+    1. [=Assert=]: the type of |outputs| is {{MLNamedOperands}}.
+    1. If |outputs| is empty, then [=exception/throw=] a {{TypeError}}.
+    1. [=map/For each=] |element| in |outputs|:
+        1. [=Assert=]: the type of |element|.key is [=string=].
+        1. If |element|.key is empty, then [=exception/throw=] a {{TypeError}}.
+        1. [=Assert=]: the type of |element|.value is {{MLOperand}}.
+    1. If any of the following sub-steps fail, [=exception/throw=] an "{{OperationError}}" {{DOMException}}.
+        1. Let |graph| be a new {{MLGraph}}:
+            1. Set |graph|.{{MLGraph/[[context]]}} to [=this=].{{MLGraphBuilder/[[context]]}}.
+        1. Make a request to the underlying platform to:
+            1. Connect |graph| to a new [=implementation-defined=] graph implementation |graphImpl| given |graph|.
+            1. Store a reference to |graphImpl| in |graph|.{{MLGraph/[[implementation]]}}.
+        1. Make a request to the underlying platform to initialize the graph:
+            1. [=map/For each=] |operand| in |outputs|:
+                1. If <a>validating MLOperand</a> given |operand| and [=this=] returns false, then [=exception/throw=] a {{TypeError}}.
+                1. If |operand| was created as an input by the underlying platform:
+                    1. If |operand|.{{MLOperand/[[name]]}}] is not unique for |graphImpl|, then [=exception/throw=] a {{TypeError}}.
+                    1. Add |operand|.{{MLOperand/[[descriptor]]}} to |graph|.{{MLGraph/[[inputDescriptors]]}}[|operand|.{{MLOperand/[[name]]}}].
+                1. If |operand| was created as a constant by the underlying platform:
+                    1. Implementations MAY preprocess and optimize the tensor data of |operand| for the underlying platform.
+                1. Register |operand|.{{MLOperand/[[operand]]}} in |graphImpl| as graph output.
+                1. Register |operand|.{{MLOperand/[[operator]]}} to |graphImpl|.
+    1. Return |graph|.
+  </div>
+</details>
+
+### cast ### {#api-mlgraphbuilder-cast}
+Cast each element in the input tensor to the target data type.
```

Comment:
(raised by @wacky6 in Chromium [CL-]() review)

Casting to/from fp <-> int or fp is well understood. But what about the casting behavior from fp / int <-> uint?

Would -1.23 fp32 cast to 0 or 255 uint8?

The spec PR mentions cast is implementation defined, which isn't ideal. We should at least provide what caller should expect.

Reply:
> what about the casting behavior from fp / int <-> uint?

**Ningxin**: That's a trickier case for conformance, casting from a wider range into a narrower range, different hardware gives differing results. On CPU, using SSE vs the classic FPU could return different results. On GPU, you could get different results depending on whether your GPU supported typed UAV's or native structured UAV's. On NPU, I don't even know yet.

> Would -1.23 fp32 cast to 0 or 255 uint8?

I can say that locally for -1.0f -> uint8, I get 255 on CPU via C++ `static_cast<uint8_t>`, but I get 0 on my GPU (because -1.0f is mapped to int32_t -1, which then clamps to `[0,255]` when written out to the typed UAV, since uint8_t is not a natively supported type within HLSL). Then for -1.0f -> uint16_t, it's a similar story, 0xFFFF on CPU but 0 on GPU. Though, if I tried this on a GPU with `D3D12_FEATURE_DATA_D3D12_OPTIONS4::Native16BitShaderOpsSupported` true, then I might well get 0xFFFF instead.

So, if we require an exact answer in the spec for negative float -> uint cases, then we'd need some intermediate casts. Surprisingly though, this issue evidently hasn't come up so far in the DML EP. Want to open an issue for it?

Reply:
Type casting from floating-point to signed/unsigned integer is a complex process, and one without clear industry standard b/c the casting will be both accuracy-lossy and range-dependent. It is generally understood simply as an "undefined" behavior where the outcome may depend on a number of runtime factors including the support in the hardware. This is part of the reason why no other framework attempts to define it concretely to-date and leave part of it as implementation-dependent. The same situation is applied to WebNN.

Reply:
If the behavior depends on different hardware internals, would it cause some finger-printing issues? Might be worth tracking it by a separate issue. WDYT?

Reply:
This definitely merits its own issue for discussion. A possible resolution is an amendment to the security and/or privacy considerations. That is non-blocking for this PR.

@huningxin please open a new issue, link it here, then mark this conversation resolved.

Reply:
Agreed.